include/hydrogen/meta/MetaUtilities.hpp

```diff
@@ -127,6 +131,24 @@ template <typename EnumT, EnumT A>
 struct EnumSame<EnumT,A,A> : std::true_type {};
 
 ///@}
+/** @brief Miscellaneous metafunctions */
+///@{
 
+/** @brief Metafunction for determining whether all types match a
+ *         given type.
+ */
+template <typename... Ts> struct AllMatch : std::false_type {};
+
+// Base
+template <typename T>
+struct AllMatch<T> : std::true_type {};
```

Comment:
Won't `AllMatch` always resolve to `std::false_type` if it has an even number of template args? I think this fixes it:
```suggestion
/** @brief Metafunction for determining whether all types match a
 *         given type.
 */
template <typename... Ts> struct AllMatch : std::true_type {};
```

Reply:
The first line is the general template. Technically, it doesn't have to even be defined. Basically it only covers one case: `AllMatch<>`, which really should be an error (because you're asking "Does my list of nothing match a type I didn't tell you?", which makes no sense) -- basically, at least 1 argument is required.

It would make more sense to define the predicate as:
```cpp
// General metafunction.
template <typename MatchType, typename... TestTypes...>
struct AllMatch;

// Base case:
template <typename MatchType, typename TestType>
struct AllMatch<MatchType, TestType> : IsSame<MatchType, TestType> {};

// Recursive call
template <typename MatchType, typename TestType, typename... Rest>
struct AllMatch<MatchType, TestType, Rest...>
  : And<IsSame<MatchType, TestType>, AllMatch<MatchType, Rest...>>
{};
```
I'll recode this and write up a "static unit test" demonstrating that it works correctly.

Reply:
Oh, I see now it's correct. I thought it was `AllMatch<Us...>` in the recursive step.

Personally, I would think that `AllMatch<>` would match to `std::true_type` since there are no mismatches (similar to `std::all_off`), but it's kind of splitting hairs.

Reply:
0 types should be an error (`AllMatch<>`), and 1 type should be `true`, i.e. `AllMatch<MatchType>`. Things like `all_off` require a predicate and not even providing `MatchType` is analogous to not providing a predicate to that function, which should be a compiler error. I'm updating the implementation to match this view.